houston/ardu/sandbox.py

```diff
@@ -102,59 +115,187 @@ def _start(self,
         protocol = 'tcp'
         port = 5760
-        ip = str(self.bugzoo.containers.ip_address(self.container))
+        ip = str(bzc.ip_address(self.container))
         url = "{}:{}:{}".format(protocol, ip, port)
+
         dummy_connection = mavutil.mavlink_connection(url)
         time.sleep(10)
         dummy_connection.close()
         time.sleep(5)
-        self.__connection = dronekit.connect(url, wait_ready=True)
+        event = threading.Event()
 
-        # wait until vehicle is ready to test
-        self.__connection.wait_ready('autopilot_version')
+        def stdout(m: MAVLinkMessage):
+            name = m.name
+            message = m.message
+            if name == "STATUSTEXT" and \
+                    "EKF2 IMU0 Origin set to GPS" in str(message.text):
+                logger.debug("Vehicle is ready")
+                event.set()
 
+        self.__connection = MAVLinkConnection(url, {'update': self.update,
+                                                    'std': stdout})
+        event.wait(10)
+        self.__connection.remove_hook('std')
+
+        # FIXME add timeout
         # wait for longitude and latitude to match their expected values, and
         # for the system to match the expected `armable` state.
-        initial_lon = mission.initial_state['longitude']
-        initial_lat = mission.initial_state['latitude']
-        v = self.system.variable
+        initial_lon = self.state_initial['longitude']
+        initial_lat = self.state_initial['latitude']
+        initial_armable = self.state_initial['armable']
+        v = self.state_initial.__class__.variables
         while True:
-            observed = self.observe(0.0)
-            ready_lon = v('longitude').eq(initial_lon, observed['longitude'])
-            ready_lat = v('latitude').eq(initial_lat, observed['latitude'])
-            ready_armable = \
-                observed['armable'] == mission.initial_state['armable']
+            # self.observe()
+            ready_lon = v['longitude'].eq(initial_lon, self.state['longitude'])
+            ready_lat = v['latitude'].eq(initial_lat, self.state['latitude'])
+            ready_armable = self.state['armable'] == initial_armable
             if ready_lon and ready_lat and ready_armable:
                 break
             time.sleep(0.05)
 
-        if not self._post_connection_setup():
-            print("Post connection setup failed!")
+        if not self._on_connected():
+            raise Exception("post-connection setup failed.")  # FIXME better exception  # noqa: pycodestyle
 
         # wait until the vehicle is in GUIDED mode
         # TODO: add timeout
         guided_mode = dronekit.VehicleMode('GUIDED')
-        self.__connection.mode = guided_mode
-        while self.__connection.mode != guided_mode:
+        self.vehicle.mode = guided_mode
+        while self.vehicle.mode != guided_mode:
             time.sleep(0.05)
+        logger.debug("Mode done")
 
-    def _stop(self) -> None:
```

Comment:
The method was named `_stop` since it shouldn't be considered to be part of the class's public interface. I'm not too fussed about it, though.

Reply:
I have no idea how this got there. It was unintended.